Document the use of mandatory macro parameters and add named macro tokens (#1212)
galaxyproject/planemo: docs/_writing_macros.rst

           <yield />
         </param>
       </when>
       <when value="history">
         <param name="ownFile" type="data" format="fasta" metadata_name="dbkey" label="Select the reference genome" />
       </when>  <!-- history -->
     </conditional>  <!-- refGenomeSource -->
   </xml>

Notice the yield statement in lieu of an options declaration. This allows the nested options element to be declared when expanding the macro:

The following expand declarations have replaced the original conditional elements.

.. code-block:: xml

   <expand macro="refGenomeSourceConditional">
     <options from_data_table="tophat_indexes">
       <filter type="sort_by" column="2"/>
       <validator type="no_options" message="No genomes are available for the selected input dataset"/>
     </options>
   </expand>

.. code-block:: xml

   <expand macro="refGenomeSourceConditional">
     <options from_data_table="tophat2_indexes">
       <filter type="sort_by" column="2"/>
       <validator type="no_options" message="No genomes are available for the selected input dataset"/>
     </options>
   </expand>

From Galaxy release 22.01 named yields are supported. That is, if the macro contains

> ``<yield name="xyz"/>`` it is replaced by the content of the ``token`` tag with the same name.

Reviewer: ```suggestion
``<yield name="xyz"/>`` it is replaced by the content of the ``token`` element with the same name.
```

> Token tags need to be direct children of the ``expand`` element. This is useful if different

Reviewer: ```suggestion
Token elements need to be direct children of the ``expand`` element. This is useful if different
```

.. code-block:: xml
   :emphasize-lines: 6,10,14

   <xml name="named_yields_example">
     <conditional>
       <param type="select">
         <option value="a">A</option>
         <option value="b">B</option>
         <yield name="more_options"/>
       </param>
       <when value="a">
         <param name="aselect" type="select">
           <yield />
         </param>
       </when>
       <when value="b"/>
       <yield name="more_whens">
     </conditional>
   </xml>

and expanding the macro in the following way:

.. code-block:: xml
   :emphasize-lines: 2,5,8,12

   <expand macro="named_yields_example">
     <token name="more_options">
       <option value="c">C</option>
     </token>
     <token name="more_whens">
       <when value="c">
         <param type="select">
           <yield />
         </param>
       </when>
     </token>
     <options from_data_table="tophat2_indexes">
       <filter type="sort_by" column="2"/>
       <validator type="no_options" message="No genomes are available for the selected input dataset"/>
     </options>
   </expand>

we get the following expanded definition:

.. code-block:: xml

>    :emphasize-lines: 6,9-14,17-24

Reviewer: ```suggestion
   :emphasize-lines: 6,10-13,17-24
```

   <xml name="named_yields_example">
     <conditional>
       <param type="select">
         <option value="a">A</option>
         <option value="b">B</option>
         <option value="c">C</option>
       </param>
       <when value="a">
         <param name="aselect" type="select">
           <options from_data_table="tophat2_indexes">
             <filter type="sort_by" column="2"/>
             <validator type="no_options" message="No genomes are available for the selected input dataset"/>
           </options>
         </param>
       </when>
       <when value="b"/>
       <when value="c">
         <param type="select">
           <options from_data_table="tophat2_indexes">
             <filter type="sort_by" column="2"/>
             <validator type="no_options" message="No genomes are available for the selected input dataset"/>
           </options>
         </param>
       </when>
     </conditional>
   </xml>

Named yields are replaced in the order of the tokens defined in the ``expand``

> tag. Unamed yields are replaced after all named tokens have been replaced (by

Reviewer: ```suggestion
tag. Unnamed yields are replaced after all named tokens have been replaced (by
```

the non-token child elements of the expand tag). If there are named
yields that have no corresponding token, then they are treated like unnamed

> yields. Note that, unnamed and named tokens can be used multiple times in a

Reviewer: ```suggestion
yields. Note that unnamed and named tokens can be used multiple times in a
```

> macro, then each occurence is preplaced by the corresponding content definedin

Reviewer: ```suggestion
macro, then each occurrence is replaced by the corresponding content defined in
```

the ``expand``.

Further, note that the order of the replacements offers some possibilities
to achieve recursion-like replacements, since a token may contain further named
or unnamed ``yield`` tags (see for instance the ``yield`` tag contained in the
named token ``more_whens``).

-----------------------------------------
Parameterizing XML Macros (with tokens)
-----------------------------------------

> In addition to using ``yield`` blocks, there is another way to parameterize

Reviewer: ```suggestion
In addition to using ``yield`` blocks, there is another way to parametrize
```

pass values for those parameters and produce varying results.

.. code-block:: xml

   <inputs>
       <expand macro="color" varname="myvar" default_color="#ff0000" />
       <expand macro="color" varname="c2" default_color="#0000ff" label="Choose a different color" />
   </inputs>

The attributes passed to the macro definition will be filled in (or defaults used if not provided).
Effectively this yields:

.. code-block:: xml

   <inputs>
       <param name="myvar" type="color" label="Pick a color" value="#ff0000" />
       <param name="c2" type="color" label="Choose a different color" value="#0000ff" />
   </inputs>

Macro tokens can be used in the text content of tags, attribute values, and (with a
little trick also in attribute names). The problem is that the default delimiting character of
macro tokens is ``@`` and the XML must still be valid before processing the macros (and ``@``
is invalid in attribute names). Luckily the delimiting character(s) can be changed by adding
``token_quote`` to the macro definition:

.. code-block:: xml

   <macros>
     <xml name="color" tokens="attr,attr_value" token_quote="__" token_label="label">

>          <param __VAR__="__ATTR_VALUE__" label="__LABEL__"/>

Reviewer: ```suggestion
         <param __ATTR__="__ATTR_VALUE__" label="__LABEL__"/>
```

          </options>
        </expand>

From Galaxy release 22.01 named yields are supported. That is, if the macro contains
``<yield name="xyz"/>`` it is replaced by the content of the ``token` tag with the same name.

> Token tags need to be direct children of the ``expand`` element.

Reviewer 2: ```suggestion
Token tags need to be direct children of the ``expand`` element.
This is useful when you would like to compose options of a select list, as shown in the example below.
```

> 

Reviewer 2: Can we have the example be something like:
Given the following macro:
...
and expanding the macro in the following way:
...
we get the following expanded definition:
...
?

.. code-block:: xml

    <xml name="named_yields_example">
      <conditional>
        <param type="select">
          <option value="a">A</option>
          <option value="b">B</option>
          <yield name="more_options"/>
        </param>
        <when value="a">
          <param type="select">
            <yield />
          </param>
        </when>
        <when value="b">

>           <param ... />

Reviewer 2: I think it helps to have this be somewhat valid XML.

        </when>
        <yield name="more_whens">
      </conditional>
    </xml>

.. code-block:: xml

>         <expand macro="named_yields_example">

Reviewer 2: The indentiation looks weird to me, why not start on the left side ?

          <token name="more_options">
            <option value="c">C</option>
          </token>
          <token name="more_whens">
            <when value="c">
              <param type="select">
                <yield />
              </param>
            </when>
          </token>

>           <options from_data_table="tophat2_indexes">
>             <filter type="sort_by" column="2"/>
>             <validator type="no_options" message="No genomes are available for the selected input dataset"/>
>           </options>

Reviewer 2: ```suggestion
```
This is unrelated to the macro example, right ?

Author: The unnamed yield is replaced by this.

Reviewer 2: Oh, I have not understood this. Why is that ? How do I read this as a tool author ?

Author: Essentially as before: The state was that the content of the expand tag is inserted at the place(s) of the yield tags in the macro. 

In order to keep backward compatibility I implemented it such still all content of the expand minus the token tags and their content is used for the replacement.

Reviewer 2: Ah, I see, yeah, that makes sense.

-----------------------------------------

In addition to using ``yield`` blocks, there is another way to parameterize
macros through the use of specifying 

> - required parameters as comma separated list of parameter names using the

Reviewer: ```suggestion
- required parameters as comma-separated list of parameter names using the
```

``tokens`` attribute of the macro, e.g. ``tokens="foo,bar"`` and then using
@FOO@ and @BAR@ anywhere in the macro 

> - optional parameters as ``token_xyz`` attributes on the macro 

Reviewer: ```suggestion
- optional parameters as ``token_xyz="default_value"`` attributes of the ``xml`` 
```

> definition, and then using ``@XYZ@`` anywhere within the XML.

Reviewer: ```suggestion
  element, and then using ``@XYZ@`` anywhere inside the macro definition.
```
